### PR TITLE
[7156] Human Readable Controller Configs

### PR DESCRIPTION
**Pull request by suddendust**
## Description
#7156
This PR deprecates certain controller configurations  in favour of new configs that accept human-readable period strings.
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes
Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes
Does this PR otherwise need attention when creating release notes?
* [x] Yes
## Release Notes
The following configurations have been updated. Only those configs have been updated for which we expect valid values to be in minutes or hours (that is, if we only expect a configuration to be supplied in seconds, such configurations haven't been changed). If both configs are present, the new one is picked. If both configs are present but the new config uses an incorrect representation of period string, an exception is thrown (it does not fallback to the old configuration).
| Deprecated Config      | New Config 
| ----------- | ----------- |
|`controller.retention.frequencyInSeconds` |`controller.retention.frequencyPeriod` 
|`controller.offline.segment.interval.checker.frequencyInSeconds` |`controller.offline.segment.interval.checker.frequencyPeriod` |
|`controller.realtime.segment.validation.frequencyInSeconds` |`controller.realtime.segment.validation.frequencyPeriod` |
|`controller.broker.resource.validation.frequencyInSeconds` |`controller.broker.resource.validation.frequencyPeriod` |
|`controller.statuschecker.waitForPushTimeInSeconds` | `controller.statuschecker.waitForPushTimePeriod`|
|`controller.statuschecker.frequencyInSeconds` |`controller.statuschecker.frequencyPeriod` |
|`controller.task.frequencyInSeconds` |`controller.task.frequencyPeriod` |
|`controller.minion.instances.cleanup.task.frequencyInSeconds` |`controller.minion.instances.cleanup.task.frequencyPeriod` |
|`controller.minion.instances.cleanup.task.minOfflineTimeBeforeDeletionSeconds` |`controller.minion.instances.cleanup.task.minOfflineTimeBeforeDeletionPeriod` |
|`controller.minion.task.metrics.emitter.frequencyInSeconds` |`controller.minion.task.metrics.emitter.frequencyPeriod` |
|`controller.segment.relocator.frequencyInSeconds` |`controller.segment.relocator.frequencyPeriod` |
|`controller.segment.level.validation.intervalInSeconds` |`controller.segment.level.validation.intervalPeriod` |

Some examples of updated configurations:
```
controller.minion.task.metrics.emitter.frequencyInSeconds=3h20m
controller.segment.level.validation.intervalPeriod=5m
```
## Documentation
Final PR for docs update will be raised once this PR is approved.

### CODE REVIEW

**Comment by suddendust** [comment]
I can't add labels for some reason. This PR needs the "release note" label.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 1]
```diff
@@ -148,15 +205,23 @@ private static long getRandomInitialDelayInSeconds() {
...
+    private static final int DEFAULT_MINION_INSTANCES_CLEANUP_TASK_MIN_OFFLINE_TIME_BEFORE_DELETION_IN_SECONDS =
+        60 * 60; // 1 Hour.
 
     private static final int DEFAULT_SEGMENT_LEVEL_VALIDATION_INTERVAL_IN_SECONDS = 24 * 60 * 60;
     private static final int DEFAULT_SEGMENT_RELOCATOR_FREQUENCY_IN_SECONDS = 60 * 60;
   }
 
+  @Deprecated
   private static final String SERVER_ADMIN_REQUEST_TIMEOUT_SECONDS = "server.request.timeoutSeconds";
+  private static final String SERVER_ADMIN_REQUEST_TIMEOUT_PERIOD = "server.request.timeoutPeriod";
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:217]: This one should not be changed. We don't expect people to configure this value in minutes or hours.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 2]
```diff
@@ -148,15 +205,23 @@ private static long getRandomInitialDelayInSeconds() {
...
+
+  @Deprecated
   private static final String SEGMENT_COMMIT_TIMEOUT_SECONDS = "controller.realtime.segment.commit.timeoutSeconds";
+  private static final String SEGMENT_COMMIT_TIMEOUT_PERIOD = "controller.realtime.segment.commit.timeoutPeriod";
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:221]: This one should not be changed

File: pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    ControllerConf conf = new ControllerConf(controllerConfig);
+    //execution and assertion
+    assertOnDurations(conf, PERIOD_DURATION_IN_SECONDS);
+  }
+
+  /**
+   * When valid unit config and invalid period config are specified, then an {@link IllegalArgumentException} is thrown (valid unit
+   * config does not override invalid period config)
+   */
+  @Test(expectedExceptions = IllegalArgumentException.class)
```
Comment (mcvsubbu) [inline pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java:105]: Expected exceptions are good, but I prefer fielding specific exceptions at the specific points where we expect them. That makes it easy to add more code to the test, or re-factor things, etc. and not worry about the test passing if we accidentally get IllegalArgumentException some place else.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 4]
```diff
@@ -341,8 +406,9 @@ public String getDataDir() {
   }
 
   public int getSegmentCommitTimeoutSeconds() {
```
Comment (jackjlli) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:408]: I feel that these method names also need to be changed, or have a new method here and mark the old one as deprecated. 
The behavior should be that the new one should be used if both are specified or only the new one be specified. If not, use the old one.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 5]
```diff
@@ -428,15 +496,20 @@ public void setRetentionControllerFrequencyInSeconds(int retentionFrequencyInSec
   }
 
   /**
-   * Returns the config value for controller.offline.segment.interval.checker.frequencyInSeconds if it exists.
-   * If it doesn't exist, returns the segment level validation interval. This is done in order to retain the current behavior,
-   * wherein the offline validation tasks were done at segment level validation interval frequency
-   * The default value is the new DEFAULT_OFFLINE_SEGMENT_INTERVAL_CHECKER_FREQUENCY_IN_SECONDS
+   * Returns the config value for controller.offline.segment.interval.checker.frequencyInSeconds if
```
Comment (jackjlli) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:499]: We should check the new one first and then the old one, right?

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 1 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:217]: Yes makes sense. My thought process was that the configuration should move towards taking everything as human-readable strings and converting it to seconds (to keep it generic, basically if user wanted to keep the timeout as 2m, he should be able to).

File: pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java (hunk 3 above)
Comment (suddendust) [reply pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java:105]: That's a valid point @mcvsubbu. I just went with the simplest approach as the test was very trivial, and you pretty much can point where exactly the exception was thrown. Will change it.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 4 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:408]: "The behavior should be that the new one should be used if both are specified or only the new one be specified" - Isn't it what it is doing @jackjlli? Basically if the new config is absent, it falls-back to the older one. But method still does what it says - `getSegmentCommitTimeoutSeconds`.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 5 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:499]: Missed this, will update the docs.

**Comment by suddendust** [comment]
Thanks for the review everyone, I will incorporate the suggested changes.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 4 above)
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:408]: Right, I wrote this one because I saw some of the java docs for the methods have been adjusted but it talks about the other way around. So mention it here.

**Comment by suddendust** [comment]
Addressed earlier comments

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 6]
```diff
@@ -74,60 +76,115 @@
...
+
     // frequency configs
+    @Deprecated
     public static final String RETENTION_MANAGER_FREQUENCY_IN_SECONDS = "controller.retention.frequencyInSeconds";
+    public static final String RETENTION_MANAGER_FREQUENCY_PERIOD = "controller.retention.frequencyPeriod";
     @Deprecated
     // The ValidationManager has been split up into 3 separate tasks, each having their own frequency config settings
     public static final String DEPRECATED_VALIDATION_MANAGER_FREQUENCY_IN_SECONDS =
         "controller.validation.frequencyInSeconds";
+    @Deprecated
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:88]: Another (separate) task can be to remove the `DEPRECATED_VALIDATION_MANAGER_FREQUENCY_IN_SECONDS` config and related code. I believe this has been deprecated for a few releases now, we can remove it.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 7]
```diff
@@ -74,60 +76,115 @@
...
         "controller.broker.resource.validation.initialDelayInSeconds";
+    public static final String BROKER_RESOURCE_VALIDATION_INITIAL_DELAY_PERIOD =
+        "controller.broker.resource.validation.initialDelayPeriod";
+    @Deprecated
     public static final String STATUS_CHECKER_FREQUENCY_IN_SECONDS = "controller.statuschecker.frequencyInSeconds";
+    public static final String STATUS_CHECKER_FREQUENCY_PERIOD = "controller.statuschecker.frequencyPeriod";
+    @Deprecated
     public static final String STATUS_CHECKER_WAIT_FOR_PUSH_TIME_IN_SECONDS =
         "controller.statuschecker.waitForPushTimeInSeconds";
+    public static final String STATUS_CHECKER_WAIT_FOR_PUSH_TIME_PERIOD =
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:119]: This config is (and please add this doc here) to make sure that a recently pushed segment is not accidentally flagged as bad just because it does not reflect in external view as yet. So, this time should be in seconds all the time. Please revert this change.

**Review by mcvsubbu — COMMENTED** [review]
Please make sure that you change only those that actually need to be in minutes or hours. If something needs to be in seconds (or even lower) chances are that the default values will work, and hardly anyone will change it. So, ew don't need to add extra code to introduce new config and deprecate old one, etc. Not to mention someone setting a very high value and maybe causing other things to break

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 8]
```diff
@@ -568,17 +666,21 @@ public void setDeletedSegmentsRetentionInDays(int retentionInDays) {
   }
 
   public int getTaskManagerFrequencyInSeconds() {
-    return getProperty(ControllerPeriodicTasksConf.TASK_MANAGER_FREQUENCY_IN_SECONDS,
-        ControllerPeriodicTasksConf.DEFAULT_TASK_MANAGER_FREQUENCY_IN_SECONDS);
+    return Optional.ofNullable(getProperty(ControllerPeriodicTasksConf.TASK_MANAGER_FREQUENCY_PERIOD))
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:669]: Does this work if value is set to "-1" in the new config (that is the default value).

**Comment by mcvsubbu** [comment]
Alsobe sure to document these in https://docs.pinot.apache.org/configuration-reference/controller

I guess the convention is to keep the old config and mark them as deprecated in the doc, and add the new config.  Be sure to add an example in the "Description" column.

**Comment by suddendust** [comment]
> Please make sure that you change only those that actually need to be in minutes or hours. If something needs to be in seconds (or even lower) chances are that the default values will work, and hardly anyone will change it. So, ew don't need to add extra code to introduce new config and deprecate old one, etc. Not to mention someone setting a very high value and maybe causing other things to break.

Yes @mcvsubbu actually I was my thought process along the lines of keeping the configuration consistent - That is, even if we always expect a config in seconds, and the user say gives 2m as the config, we convert it to 120s and use it. But your points show this can be fragile. I will remove such configs.

**Comment by suddendust** [comment]
> Alsobe sure to document these in https://docs.pinot.apache.org/configuration-reference/controller
> 
> I guess the convention is to keep the old config and mark them as deprecated in the doc, and add the new config. Be sure to add an example in the "Description" column.

This is a WIP. I was actually confident that some of these new configurations would be rolled-back, so I waited till a review was done.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 9]
```diff
@@ -74,60 +76,115 @@
   }
 
   public static class ControllerPeriodicTasksConf {
+
     // frequency configs
+    @Deprecated
     public static final String RETENTION_MANAGER_FREQUENCY_IN_SECONDS = "controller.retention.frequencyInSeconds";
```
Comment (jackjlli) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:82]: I like the idea of adding `DEPRECATED_` prefix to the ones that are about to deprecated in the next release. Maybe we should do that in this PR as well.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 7 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:119]: @mcvsubbu this config's default value in [docs](https://docs.pinot.apache.org/configuration-reference/controller#:~:text=controller.statuschecker.waitForPushTimeInSeconds) is 10 minutes. So can't we expect that the user might want to supply a different value in minutes again?

Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:119]: You are right. I was mixed up with another config perhaps. Thanks for the diligence

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 9 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java]: Done.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 8 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java]: Handled this in the latest commit and added a UT to catch this. This wasn't handled earlier.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 6 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:87]: All deprecated configs have been prefixed with `DEPRECATED.` I will create a new ticket to do this in other places.

Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:87]: I was suggesting to just remove `DEPRECATED_VALIDATION_MANAGER_FREQUENCY_IN_SECONDS`. It has been there since 0.5.0.  I understand this can mean some more code to remove. If that is too much, then you can choose to add a TODO there, saying that this config has been deprecated as of 0.5.0, so must be removed. Thanks

**Review by mcvsubbu — COMMENTED** [review]
also, in all the changes that you make, if you can prefix the configs with a comment like this:
// Deprecated as of 0.8.0
Then, it is easy for someone to remove it a few releases down the road. 

Other than these minor comments, LGTM

Thanks for your contribution

**Comment by suddendust** [comment]
> also, in all the changes that you make, if you can prefix the configs with a comment like this:
> // Deprecated as of 0.8.0
> Then, it is easy for someone to remove it a few releases down the road.
> 
> Other than these minor comments, LGTM
> 
> Thanks for your contribution

Hey sure @mcvsubbu it is a pleasure to contribute to Pinot :) I will make these changes and raise another PR after updating the documentation. Thanks for your thorough review.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableSegmentUploadIntegrationTest.java [hunk 10]
```diff
@@ -152,8 +152,8 @@ protected void startController()
     Map<String, Object> controllerConfig = getDefaultControllerConfiguration();
     // Perform realtime segment validation every second with 1 second initial delay.
     controllerConfig
-        .put(ControllerConf.ControllerPeriodicTasksConf.REALTIME_SEGMENT_VALIDATION_FREQUENCY_IN_SECONDS, 1);
-    controllerConfig.put(ControllerConf.ControllerPeriodicTasksConf.SEGMENT_LEVEL_VALIDATION_INTERVAL_IN_SECONDS, 1);
+        .put(ControllerConf.ControllerPeriodicTasksConf.DEPRECATED_REALTIME_SEGMENT_VALIDATION_FREQUENCY_IN_SECONDS, 1);
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableSegmentUploadIntegrationTest.java:155]: The deprecated configs should only be used in the getter/setter method for better cleanup purpose. Maybe we should start changing these config to the ones you introduced in this PR?

**Review by jackjlli — APPROVED** [review]
Minor but LGTM. Thanks @suddendust for your contribution!

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableSegmentUploadIntegrationTest.java (hunk 10 above)
Comment (suddendust) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableSegmentUploadIntegrationTest.java:155]: Yes @jackjlli we need to stop using these deprecated configs in our tests. Can I raise another PR for this? I think they are used at too many places and changing it in this PR will add another round of review. Can raise it immediately after this one.

**Comment by suddendust** [comment]
PR for docs update: https://github.com/pinot-contrib/pinot-docs/pull/47

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 6 above)
Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:89]: Can you either remove `DEPRECATED_VALIDATION_MANAGER_FREQUENCY_IN_SECONDS` or add a comment before line 80 saying it has been deprecated since 0.5.0

File: pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+   * corresponding deprecated configuration.
+   */
+  @Test
+  public void invalidNewConfigShouldThrowExceptionOnReadWithoutFallbackToCorrespondingValidDeprecatedConfig() {
+    //setup
+    Map<String, Object> controllerConfig = new HashMap<>();
+    int durationInSeconds = getRandomDurationInSeconds();
+    //all deprecated configs should be valid
+    DEPRECATED_CONFIGS.forEach(config -> controllerConfig.put(config, durationInSeconds));
+    String randomPeriodInMinutes = getRandomPeriodInMinutes();
```
Comment (mcvsubbu) [inline pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java:98]: In this case maybe it is ok, but in general, it is a good idea to set the seed for the random number up front, and print out the value of the seed (in system.out). This will make sure that if there is failure due to some random value, we can reproduce the test case by manually setting the seed to that value. Otherwise, we will face intermittent failures in tests without knowing why that happened.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java (hunk 8 above)
Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java]: I am sure I am missing something. I don't see the UT for this case?

Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java]: ah, got it.

**Comment by mcvsubbu** [comment]
Other than one minor request of adding a comment, we are good to merge this.

File: pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java (hunk 11 above)
Comment (suddendust) [reply pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java:98]: Ah this is a good learning. Will incorporate this change.

Comment (suddendust) [reply pinot-controller/src/test/java/org/apache/pinot/controller/ControllerConfTest.java:98]: While this definitely is needed for reproducing failing tests, in this case, simply printing the test configuration gives enough information to reproduce a failing test. So I have simply printing the config for any failing test.

**Comment by suddendust** [comment]
> Other than one minor request of adding a comment, we are good to merge this.

Addressed the changes.

**Review by mcvsubbu — APPROVED** [review]
Thanks again. Good, clean contribution.

**Comment by mcvsubbu** [comment]
Marking this for release notes, as well as backward incompat (since one config has been removed. This config was deprecated in 0.5.0 version)
